geom/marshal_unmarshal_test.go

```diff
@@ -0,0 +1,245 @@
+		"GEOMETRYCOLLECTION Z (POINT Z (1 2 3),MULTIPOLYGON Z EMPTY)",
+		"GEOMETRYCOLLECTION Z (POINT Z (1 2 3),POLYGON Z EMPTY)",
+		"GEOMETRYCOLLECTION Z (POINT Z (1 2 3),MULTILINESTRING Z EMPTY)",
+		"GEOMETRYCOLLECTION Z (POINT Z (1 2 3),GEOMETRYCOLLECTION Z EMPTY)",
+
+		"GEOMETRYCOLLECTION M EMPTY",
+		"GEOMETRYCOLLECTION M (GEOMETRYCOLLECTION M EMPTY)",
+		"GEOMETRYCOLLECTION M (POINT M EMPTY)",
+		"GEOMETRYCOLLECTION M (LINESTRING M EMPTY)",
+		"GEOMETRYCOLLECTION M (POLYGON M EMPTY)",
+		"GEOMETRYCOLLECTION M (MULTIPOINT M EMPTY)",
+		"GEOMETRYCOLLECTION M (MULTILINESTRING M EMPTY)",
+		"GEOMETRYCOLLECTION M (MULTIPOLYGON M EMPTY)",
+		"GEOMETRYCOLLECTION M (LINESTRING M (0 0 3,1 1 4))",
+		"GEOMETRYCOLLECTION M (POINT M (4 6 1),LINESTRING M (4 6 5,7 10 11))",
+		"GEOMETRYCOLLECTION M (POINT M (1 2 3),MULTIPOLYGON M EMPTY)",
+		"GEOMETRYCOLLECTION M (POINT M (1 2 3),POLYGON M EMPTY)",
+		"GEOMETRYCOLLECTION M (POINT M (1 2 3),MULTILINESTRING M EMPTY)",
+		"GEOMETRYCOLLECTION M (POINT M (1 2 3),GEOMETRYCOLLECTION M EMPTY)",
+
+		"GEOMETRYCOLLECTION ZM EMPTY",
+		"GEOMETRYCOLLECTION ZM (GEOMETRYCOLLECTION ZM EMPTY)",
+		"GEOMETRYCOLLECTION ZM (POINT ZM EMPTY)",
+		"GEOMETRYCOLLECTION ZM (LINESTRING ZM EMPTY)",
+		"GEOMETRYCOLLECTION ZM (POLYGON ZM EMPTY)",
+		"GEOMETRYCOLLECTION ZM (MULTIPOINT ZM EMPTY)",
+		"GEOMETRYCOLLECTION ZM (MULTILINESTRING ZM EMPTY)",
+		"GEOMETRYCOLLECTION ZM (MULTIPOLYGON ZM EMPTY)",
+		"GEOMETRYCOLLECTION ZM (LINESTRING ZM (0 0 3 1,1 1 4 2))",
+		"GEOMETRYCOLLECTION ZM (POINT ZM (4 6 1 8),LINESTRING ZM (4 6 5 7,7 10 11 0))",
+		"GEOMETRYCOLLECTION ZM (POINT ZM (1 2 3 4),MULTIPOLYGON ZM EMPTY)",
+		"GEOMETRYCOLLECTION ZM (POINT ZM (1 2 3 4),POLYGON ZM EMPTY)",
+		"GEOMETRYCOLLECTION ZM (POINT ZM (1 2 3 4),MULTILINESTRING ZM EMPTY)",
+		"GEOMETRYCOLLECTION ZM (POINT ZM (1 2 3 4),GEOMETRYCOLLECTION ZM EMPTY)",
+	}
+
+	t.Run("wkt", func(t *testing.T) {
+		for i, wkt := range wkts {
+			t.Run(strconv.Itoa(i), func(t *testing.T) {
+				got := geomFromWKT(t, wkt).AsText()
+				expectStringEq(t, got, wkt)
+			})
+		}
+	})
+
+	t.Run("wkb", func(t *testing.T) {
+		for i, wkt := range wkts {
+			t.Run(strconv.Itoa(i), func(t *testing.T) {
+				original := geomFromWKT(t, wkt)
+				var wkb bytes.Buffer
+				expectNoErr(t, original.AsBinary(&wkb))
+				reconstructed, err := geom.UnmarshalWKB(&wkb)
+				expectNoErr(t, err)
+				reconstructedWKT := reconstructed.AsText()
+				expectStringEq(t, wkt, reconstructedWKT)
+			})
+		}
+	})
+
+	t.Run("geojson", func(t *testing.T) {
+		for i, wkt := range wkts {
+			original := geomFromWKT(t, wkt)
+			if original.CoordinatesType().IsMeasured() {
+				// GeoJSON will drop Measures
+				continue
+			}
+			if original.IsEmpty() && original.CoordinatesType().Is3D() {
+				// When parsing GeoJSON, the "width" of each coordinate is used
+				// to infer XY vs XYZ geometry. If there are no coordinates,
+				// then XY is assumed.
```

Comment:
Nit: Might be good to add "Therefore GeoJSON will drop empty 3D" or similar concluding thought.

Reply:
Good call, I've now elaborated on that comment.